src/SDKs/Azure.Base/data-plane/Azure.Base/UriExtensions.cs

```diff
@@ -2,22 +2,174 @@
+            set
+            {
+                ResetUri();
+                _host = value;
+            }
+        }
+
+        public int Port
+        {
+            get => _port;
+            set
+            {
+                ResetUri();
+                _port = value;
+            }
+        }
+
+        public string Query
+        {
+            get => HasQuery ? _pathAndQuery.ToString(_queryIndex, _pathAndQuery.Length - _queryIndex) : string.Empty;
+            set
+            {
+                ResetUri();
+                if (HasQuery)
+                {
+                    _pathAndQuery.Remove(_queryIndex, _pathAndQuery.Length - _queryIndex);
+                    _queryIndex = -1;
+                }
+
+                if (!string.IsNullOrEmpty(value))
+                {
+                    _queryIndex = _pathAndQuery.Length;
+                    _pathAndQuery.Append('?');
+                    _pathAndQuery.Append(value);
+                }
+            }
+        }
+
+        public string Path
+        {
+            get => HasQuery ? _pathAndQuery.ToString(0, _queryIndex) : _pathAndQuery.ToString();
+            set
+            {
+                if (HasQuery)
+                {
+                    _pathAndQuery.Remove(0, _queryIndex);
+                    _pathAndQuery.Insert(0, value);
+                    _queryIndex = value.Length;
+                }
+                else
+                {
+                    _pathAndQuery.Remove(0, _pathAndQuery.Length);
+                    _pathAndQuery.Append(value);
+                }
+            }
+        }
+
+        private bool HasQuery => _queryIndex != -1;
+
+        public string PathAndQuery => _pathAndQuery.ToString();
+
+        public Uri Uri
         {
-            if(!string.IsNullOrEmpty(builder.Query)) {
-                builder.Query = builder.Query + "&" + name + "=" + value;
-            } 
-            else {
-                builder.Query = name + "=" + value;
+            get
+            {
+                if (_uri == null)
+                {
+                    _uri = new Uri(ToString());
+                }
+                return _uri;
+            }
+            set
+            {
+                Scheme = value.Scheme;
+                Host = value.Host;
+                Port = value.Port;
+                Path = value.AbsolutePath;
+                Query = value.Query;
+                _uri = value;
             }
         }
 
-        public static void AppendQuery(this UriBuilder builder, string name, long value)
-            => AppendQuery(builder, name, value.ToString());
+        public void AppendQuery(string name, string value)
+        {
+
+            if (!HasQuery)
+            {
+                _pathAndQuery.Append('?');
+                _queryIndex = _pathAndQuery.Length;
+            }
+            else
+            {
+                _pathAndQuery.Append('&');
+            }
+
+            _pathAndQuery.Append(name);
```

Comment:
The .NET [UriBuilder](https://github.com/Microsoft/referencesource/blob/3b1eaf5203992df69de44c783a3eda37d3d4cd10/System/net/System/uribuilder.cs#L224) has some additional functionality for escaping.   Is it safe to assume that we're good trading that for a more efficient builder?

Reply:
We would need to have escaping as well. We would have to copy parts of https://github.com/dotnet/corefx/blob/master/src/System.Private.Uri/src/System/UriHelper.cs#L131.

Reply:
Or just call Uri.EscapeUriString before constructing the Uri from our builder.

Reply:
Looks like you've chosen to do that escaping explicitly in the `ToString`, which seems like a solid decision to me.